File: algorithms/dynamic-programming/sequence-alignment/kotlin/SequenceAlignment.kt
const val GAP_COST = 4
const val MISMATCH_COST = 3

fun sequenceAlignment(s1: String, s2: String): Int {
    val m = s1.length
    val n = s2.length
    val dp = Array(m + 1) { IntArray(n + 1) }

    for (i in 0..m) dp[i][0] = i * GAP_COST
    for (j in 0..n) dp[0][j] = j * GAP_COST

    for (i in 1..m) {
        for (j in 1..n) {
            val matchCost = if (s1[i - 1] == s2[j - 1]) 0 else MISMATCH_COST
            dp[i][j] = minOf(
                dp[i - 1][j - 1] + matchCost,
                dp[i - 1][j] + GAP_COST,
                dp[i][j - 1] + GAP_COST
            )
        }
    }

    return dp[m][n]
}

fun main() {
    println(sequenceAlignment("GCCCTAGCG", "GCGCAATG")) // 18
}

File: algorithms/dynamic-programming/sequence-alignment/swift/SequenceAlignment.swift
let gapCost = 4
let mismatchCost = 3

func sequenceAlignment(_ s1: String, _ s2: String) -> Int {
    let arr1 = Array(s1)
    let arr2 = Array(s2)
    let m = arr1.count
    let n = arr2.count

    var dp = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)

    for i in 0...m { dp[i][0] = i * gapCost }
    for j in 0...n { dp[0][j] = j * gapCost }

    for i in 1...max(m, 1) {
        guard m > 0 else { break }
        for j in 1...max(n, 1) {
            guard n > 0 else { break }
            let matchCost = arr1[i - 1] == arr2[j - 1] ? 0 : mismatchCost
            dp[i][j] = min(
                min(dp[i - 1][j] + gapCost, dp[i][j - 1] + gapCost),
                dp[i - 1][j - 1] + matchCost
            )
        }
    }

    return dp[m][n]
}

print(sequenceAlignment("GCCCTAGCG", "GCGCAATG")) // 18
